cluster/main.go

```diff
@@ -0,0 +1,192 @@
+	"github.com/ipfs/ipfs-cluster/informer/numpin"
+	"github.com/ipfs/ipfs-cluster/monitor/pubsubmon"
+	"github.com/ipfs/ipfs-cluster/observations"
+	"github.com/ipfs/ipfs-cluster/pintracker/maptracker"
+	"github.com/ipfs/ipfs-cluster/pintracker/stateless"
+	host "github.com/libp2p/go-libp2p-host"
+	dht "github.com/libp2p/go-libp2p-kad-dht"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+var log = logging.Logger("tex-cluster")
+
+func InitCluster(repoPath, listenAddr string) error {
+	cfgMgr, cfgs := makeClusterConfigs()
+	err := cfgMgr.Default()
+	if err != nil {
+		return err
+	}
+
+	if listenAddr != "" {
+		addr, err := ma.NewMultiaddr(listenAddr)
+		if err != nil {
+			return err
+		}
+		cfgs.ClusterCfg.ListenAddr = addr
+	}
+
+	return cfgMgr.SaveJSON(ConfigPath(repoPath))
+}
+
+func ConfigPath(repoPath string) string {
+	return filepath.Join(repoPath, "service.json")
+}
+
+func MakeAndLoadConfigs(repoPath string) (*config.Manager, *cfgs, error) {
+	cfgMgr, cfgs := makeClusterConfigs()
+	err := cfgMgr.LoadJSONFromFile(ConfigPath(repoPath))
+	if err != nil {
+		return nil, nil, err
+	}
+	return cfgMgr, cfgs, nil
+}
+
+func ParseBootstraps(addrs []string) ([]ma.Multiaddr, error) {
+	var parsed []ma.Multiaddr
+	for _, a := range addrs {
+		p, err := ma.NewMultiaddr(a)
+		if err != nil {
+			return nil, err
+		}
+		parsed = append(parsed, p)
+	}
+	return parsed, nil
+}
+
+func Bootstrap(ctx context.Context, cluster *ipfscluster.Cluster, cons ipfscluster.Consensus, bootstraps []ma.Multiaddr) {
+	for _, bstrap := range bootstraps {
+		log.Infof("Bootstrapping to %s", bstrap)
+		err := cluster.Join(ctx, bstrap)
+		if err != nil {
+			log.Errorf("bootstrap to %s failed: %s", bstrap, err)
+		} else {
+			for _, p := range cluster.Peers(ctx) {
+				err = cons.Trust(ctx, p.ID)
+				if err != nil {
+					log.Errorf("failed to trust %s: %s", p.ID.Pretty(), err)
+				}
+			}
+		}
+	}
+}
+
+func SetupAllocation(
+	name string,
+	diskInfCfg *disk.Config,
+	numpinInfCfg *numpin.Config,
+) (ipfscluster.Informer, ipfscluster.PinAllocator, error) {
+	switch name {
+	case "disk", "disk-freespace":
+		informer, err := disk.NewInformer(diskInfCfg)
+		if err != nil {
+			return nil, nil, err
+		}
+		return informer, descendalloc.NewAllocator(), nil
+	case "disk-reposize":
+		informer, err := disk.NewInformer(diskInfCfg)
+		if err != nil {
+			return nil, nil, err
+		}
+		return informer, ascendalloc.NewAllocator(), nil
+	case "numpin", "pincount":
+		informer, err := numpin.NewInformer(numpinInfCfg)
+		if err != nil {
+			return nil, nil, err
+		}
+		return informer, ascendalloc.NewAllocator(), nil
+	default:
+		return nil, nil, fmt.Errorf("unknown allocation strategy")
+	}
+}
+
+func SetupPinTracker(
+	name string,
+	h host.Host,
+	mapCfg *maptracker.Config,
+	statelessCfg *stateless.Config,
+	peerName string,
+) (ipfscluster.PinTracker, error) {
+	switch name {
+	case "map":
```

Comment:
for the moment defaulting to this one makes most sense. We want to level up the `stateless` one.

Reply:
👌